jicofo-common/src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.kt

```diff
@@ -0,0 +1,769 @@
+                member.resetCachedRole()
+            } else {
+                logger.error("Role reset for: $occupantJid who does not exist")
+            }
+        }
+    }
+
+    /**
+     * Sets the new role for the local user in the context of this chat room.
+     */
+    private fun setLocalUserRole(newRole: MemberRole) {
+        val oldRole = role
+        role = newRole
+        if (oldRole !== newRole) {
+            eventEmitter.fireEvent { localRoleChanged(newRole, role) }
+        }
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Synchronized
+    override fun containsPresenceExtension(elementName: String, namespace: String): Boolean {
+        return (lastPresenceSent != null
+                && lastPresenceSent!!.getExtension(QName(namespace, elementName))
+                != null)
+    }
+
+    override fun grantOwnership(member: ChatRoomMember) {
+        logger.debug("Grant owner to $member")
+
+        // Have to construct the IQ manually as Smack version used here seems
+        // to be using wrong namespace(muc#owner instead of muc#admin)
+        // which does not work with the Prosody.
+        val admin = MUCAdmin()
+        admin.type = IQ.Type.set
+        admin.to = roomJid
+        val item = MUCItem(MUCAffiliation.owner, member.jid!!.asBareJid())
+        admin.addItem(item)
+        val connection = xmppProvider.xmppConnection
+        try {
+            val reply = connection.sendIqAndGetResponse(admin)
+            if (reply == null || reply.type != IQ.Type.result) {
+                // XXX FIXME throw a declared exception.
+                throw RuntimeException("Failed to grant owner: " + if (reply == null) "" else reply.toXML())
+            }
+        } catch (e: SmackException.NotConnectedException) {
+            // XXX FIXME throw a declared exception.
+            throw RuntimeException("Failed to grant owner - XMPP disconnected", e)
+        }
+    }
+
+    fun getOccupant(chatMember: ChatRoomMemberImpl): Occupant? = muc.getOccupant(chatMember.occupantJid)
+
+    /**
+     * Returns the MUCUser packet extension included in the packet or
+     * <tt>null</tt> if none.
+     *
+     * @param packet the packet that may include the MUCUser extension.
+     * @return the MUCUser found in the packet.
+     */
+    private fun getMUCUserExtension(packet: Presence?): MUCUser? {
+        return packet?.getExtension(MUCUser::class.java)
+    }
+
+    override fun setPresenceExtension(extension: ExtensionElement, remove: Boolean) {
+        var presenceToSend: Presence? = null
+        synchronized(this) {
+            if (lastPresenceSent == null) {
+                logger.error("No presence packet obtained yet")
+                return
+            }
+            var presenceUpdated = false
+
+            // Remove old
+            val old = lastPresenceSent!!.getExtension(extension.qName)
+            if (old != null) {
+                lastPresenceSent!!.removeExtension(old)
```

Comment:
Is there a way to rewrite this so Kotlin's non-null inference works for lastPresenceSent?